cloud-controller-manager/do/resources.go

```diff
@@ -0,0 +1,142 @@
+limitations under the License.
+*/
+
+package do
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/digitalocean/godo"
+	"github.com/golang/glog"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	v1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	v1lister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+)
+
+const (
+	controllerSyncPeriod     = 1 * time.Minute
+	requestTimeout           = 10 * time.Second
+	resourceTypeLoadBalancer = "load_balancer"
+)
+
+// ResourcesController ensures that DO resources are properly tagged.
+type ResourcesController struct {
+	clusterIDTag string
+	kclient      kubernetes.Interface
+	gclient      *godo.Client
+	svcLister    v1lister.ServiceLister
+}
+
+// NewResourcesController returns a new services controller responsible for
+// re-applying cluster IDs on DO load-balancers.
+func NewResourcesController(clusterIDTag string, inf v1informers.ServiceInformer,
+	k kubernetes.Interface, g *godo.Client) *ResourcesController {
+	return &ResourcesController{
+		clusterIDTag: clusterIDTag,
+		kclient:      k,
+		gclient:      g,
+		svcLister:    inf.Lister(),
+	}
+}
+
+// Run starts the resources controller. It watches over DigitalOcean resources
+// making sure the right tags are set.
+func (r *ResourcesController) Run(stopCh <-chan struct{}) {
+	ticker := time.NewTicker(controllerSyncPeriod)
+	defer ticker.Stop()
+	// Do not wait for initial tick to pass; run immediately for reduced sync
+	// latency.
+	for tickerC := ticker.C; ; {
+		if err := r.sync(); err != nil {
+			glog.Errorf("failed to sync load-balancer tags: %s", err)
+		}
+		select {
+		case <-tickerC:
+			continue
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (r *ResourcesController) sync() error {
+	// Fetch all load-balancers.
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	lbs, _, err := r.gclient.LoadBalancers.List(ctx, &godo.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list load-balancers: %s", err)
+	}
+
+	// Collect tag resources for known load balancer names (i.e., services
+	// with type=LoadBalancer.)
+	svcs, err := r.svcLister.List(labels.Everything())
+	if err != err {
+		return fmt.Errorf("failed to list services: %s", err)
+	}
+	var res []godo.Resource
+	for _, lb := range lbs {
+		for _, svc := range svcs {
+			if svc.Spec.Type == corev1.ServiceTypeLoadBalancer && cloudprovider.GetLoadBalancerName(svc) == lb.Name {
+				res = append(res, godo.Resource{
+					ID:   lb.ID,
+					Type: godo.ResourceType(resourceTypeLoadBalancer),
+				})
+			}
+		}
+	}
+
+	if len(res) == 0 {
+		return nil
+	}
+
+	// Tag collected resources with the cluster ID. If the tag does not exist
+	// (for reasons outlined below), we will create it and retry tagging again.
+	for {
+		ctx, cancel = context.WithTimeout(context.Background(), requestTimeout)
+		defer cancel()
```

Comment:
Not sure about this myself, but should this `context.WithTimeout` and `defer cancel()` be outside of this loop?

I'm unsure what the desired semantics here are. Do we want to have this loop in `sync()` run until success or timeout, or do we want each individual invocation of `sync()` to have a timeout?

Reply:
My primary intention was to bound the time spent with the requests _somehow_ so that a single sync run cannot starve subsequent ones.

Doing the timeout / cancel before vs. inside the loop defines whether we want to bound sync-overall vs. per-loop. Practically, I don't think it makes too much of a difference since the loop executes at most three requests over its lifetime span of up to two iterations (1st loop: 404 on the first tag request followed by tag creation; 2nd loop: following successful tag request). So each sync run will take ~30 seconds max (3 * 10 sec timeout plus neglectable CPU time).
Of course, that's only if the dependent APIs behave as expected. I guess we could argue that a sync-overall timeout would provide protection even for pathological cases. Not sure how realistic that might be, however.

Happy to discuss.

Reply:
Yup, just wanted to clarify the intention here. I'm inclined to suggest me move the cancel outside of the loop and rely on sync overall timeout because it means that if new LBs are created between syncs the new ones would be picked up in the next sync call.

Reply:
Yeah this loop is a bit funky, can be structured in a nicer way I think.

Reply:
Tried a different approach in https://github.com/digitalocean/digitalocean-cloud-controller-manager/pull/184, PTAL.